File: 173-binary-search-tree-iterator/binary-search-tree-iterator.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class BSTIterator {
    private Stack<TreeNode> stack;

    public BSTIterator(TreeNode root) {
        stack = new Stack<TreeNode>();
        // Initialize the stack with the leftmost path of the tree
        PushLeftNodes(root);
    }
    
    // Helper method to push all left nodes of a given node
    private void PushLeftNodes(TreeNode node) {
        while (node != null) {
            stack.Push(node);
            node = node.left;
        }
    }

    public int Next() {
        // Pop the top node from the stack
        TreeNode topNode = stack.Pop();
        // If the popped node has a right child, push its leftmost path to the stack
        if (topNode.right != null) {
            PushLeftNodes(topNode.right);
        }
        // Return the value of the node
        return topNode.val;
    }
    
    public bool HasNext() {
        // The stack being non-empty means there are still elements in the traversal
        return stack.Count > 0;
    }
}


/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.Next();
 * bool param_2 = obj.HasNext();
 */

File: 173-binary-search-tree-iterator/binary-search-tree-iterator.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 */
var BSTIterator = function(root) {
    this.nodes=[];
    this._inorderTraversal(root);
    this.index=0;
};

/**
 * @return {number}
 */
BSTIterator.prototype.next = function() {
return this.nodes[this.index++];
};

/**
 * @return {boolean}
 */
BSTIterator.prototype.hasNext = function() {
    return this.index <this.nodes.length;
};

/**
*@param {TreeNode} root
*@return {void}
 */
 BSTIterator.prototype._inorderTraversal=function(root){
    if(root){
        this._inorderTraversal(root.left);
        this.nodes.push(root.val);
        this._inorderTraversal(root.right);
    }
 };
/** 
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = new BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */